File: Heap_Sort.cpp
void heapify(int arr[], int n, int i)  
    {
        int large = i;
        int l = 2*i+1;
        int r = 2*i+2;
        
        if(l<n && arr[large]<arr[l]) large=l;
        if(r<n && arr[large]<arr[r]) large=r;
         
      
      if(large!=i){
          swap(arr[large],arr[i]);
          heapify(arr,n,large);
      }
    }

    public:
    //Function to build a Heap from array.
    void buildHeap(int arr[], int n)  
    { 
        for(int i=n/2-1;i>=0;i--)
            heapify(arr,n,i);
    }
    
    public:
    //Function to sort an array using Heap Sort.
    void heapSort(int arr[], int n)
    {
        buildHeap(arr,n);
        int size = n;
        
        for(int i=n-1;i>=0;i--){
            swap(arr[0],arr[i]);
            heapify(arr,i,0);
        }
    }

File: Heap_Sort.java
class Solution
{
    //Function to build a Heap from array.
     void buildHeap(int arr[], int n)
    {
      for (int i = n/2; i>=0; i--){
            heapify(arr, n , i );
        }  // Your code here
       
    }
 
    //Heapify function to maintain heap property.
    void heapify(int arr[], int n, int i)
    {
        int max = i;
        int left = 2*i+1;
        int right = 2*i+2;
        
        if (left<n && arr[left]> arr[max]){
            max = left;
        }
        if (right <n && arr[right] > arr[max]){
            max = right;
        }
        if (max!=i){
            //swap
            int temp = arr[i];
            arr[i] = arr[max];
            arr[max] = temp;
            heapify(arr, n, max);
        }
        // Your code here
    }
    
    //Function to sort an array using Heap Sort.
    public void heapSort(int arr[], int n)
    {
        buildHeap(arr, n);
         
        for (int i = n-1; i>0; i--){
            //swap
            int temp  = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;
            heapify (arr,i, 0);
        }
        //code here
    }
 }
